src/app/rpmostree-dbus-helpers.cxx

```diff
@@ -1524,3 +1526,166 @@ rpmostree_print_cached_update (GVariant         *cached_update,
+  g_assert (*unit_objpath);
+
+  return TRUE;
+}
+
+/* Get the property_name property of sd_unit. Returns a reference to the GVariant
+ * instance that holds the value for property_name GVariant in cache_val. */
+static gboolean
+get_sd_unit_property (GDBusConnection *connection,
+                      const char      *objpath,
+                      const char      *property_name,
+                      GVariant       **cache_val,
+                      GCancellable    *cancellable,
+                      GError         **error)
+{
+  /* Look up property_name property of systemd unit. */
+  g_autoptr(GDBusProxy) unit_obj_proxy =
+    g_dbus_proxy_new_sync (connection, G_DBUS_PROXY_FLAGS_DO_NOT_CONNECT_SIGNALS, NULL,
+                           "org.freedesktop.systemd1", objpath,
+                           "org.freedesktop.systemd1.Unit", cancellable, error);
+  if (!unit_obj_proxy)
+    return FALSE;
+
+  *cache_val =
+    g_dbus_proxy_get_cached_property (unit_obj_proxy, property_name);
+  if (!*cache_val)
+    return glnx_throw (error, "%s property not found in proxy's cache (%s)",
+                       property_name, objpath);
+
+  return TRUE;
+}
+
+/* Helper function to append docs information for sd_unit to str.
+ * Prints any errors that occur. */
+static void
+append_docs_to_str (GString          *str,
+                    GDBusConnection  *connection,
+                    char             *sd_unit,
+                    GCancellable     *cancellable)
+{
+  g_autoptr(GVariant) docs_array = NULL;
+  g_autoptr(GError) local_error = NULL;
+  const char *objpath = NULL;
+  if (!get_sd_unit_objpath (connection, "LoadUnit", g_variant_new ("(s)", sd_unit),
+                            &objpath, cancellable, &local_error))
+    g_printerr ("%s", local_error->message);
+  if (!get_sd_unit_property (connection, objpath, "Documentation", &docs_array, cancellable, &local_error))
+    {
+      g_printerr ("%s", local_error->message);
+    }
+  else if (docs_array)
+    {
+      gsize docs_len;
+      g_autofree const char **docs =
+        g_variant_get_strv (docs_array, &docs_len);
+      if (docs_len > 0)
+        {
+          g_string_append_printf (str, " at ");
+          for (guint i = 0; i < docs_len; i++)
+            g_string_append_printf (str, "%s%s", docs[i],
+                                    i < docs_len - 1 ? ", " : "\n");
+        }
+      else
+        {
+          g_string_append_printf (str, "\n");
+        }
+    }
+}
+
+/* Check whether sd_unit contains pid and return the boolean in sd_unit_contains_pid */
+static gboolean
+check_sd_unit_contains_pid (char            *sd_unit,
+                            pid_t            pid,
+                            gboolean        *sd_unit_contains_pid,
+                            GDBusConnection *connection,
+                            GCancellable    *cancellable,
+                            GError         **error)
+{
+  // Get the systemd unit associated with pid.
```

Comment:
I think we can also use the C API for this, see the call `sd_pid_get_unit (client->pid, &client->sd_unit)` in the daemon code.  But, this is fine.

Or hmm...perhaps what we should really do is check this logic on the *daemon* side and not the client.  That might just make everything simpler and avoid duplicating code.